File: app/src/main/java/omniex/nl/omniex/ui/app/splash/SplashPresenter.java
package omniex.nl.omniex.ui.app.splash;

import java.util.Objects;

import javax.inject.Inject;

import nl.omniex.omniexshopping.data.model.auth.OldToken;
import nl.omniex.omniexshopping.data.model.response.AccessTokenResponse;
import nl.omniex.omniexshopping.data.repository.TokenRepository;
import nl.omniex.omniexshopping.ui.base.BasePresenter;
import nl.omniex.omniexshopping.utils.SharedPrefUtils;
import retrofit2.Response;
import timber.log.Timber;

public class SplashPresenter extends BasePresenter<SplashView> {

    private TokenRepository mTokenRepository;
    private SharedPrefUtils mSharedPrefUtils;

    @Inject
    SplashPresenter(TokenRepository tokenRepository, SharedPrefUtils sharedPrefUtils) {
        mTokenRepository = tokenRepository;
        mSharedPrefUtils = sharedPrefUtils;
    }

    void resolveAccessToken() {
        if (mSharedPrefUtils.getAccessToken().isEmpty())
            getAccessToken();
        else
            refreshAccessToken();
    }

    private void getAccessToken() {
        addDisposable(mTokenRepository
                .getAccessToken()
                .subscribe(
                        this::handleTokenResponse,
                        error -> {
                            ifViewAttached(view -> view.onTokenError(error.getMessage()));
                            Timber.e(error);
                        })
        );
    }

    private void refreshAccessToken() {
        addDisposable(mTokenRepository
                .refreshAccessToken(new OldToken(mSharedPrefUtils.getOldToken()))
                .subscribe(
                        this::handleTokenResponse,
                        error -> {
                            ifViewAttached(view -> view.onTokenError(error.getMessage()));
                            Timber.e(error);
                        })
        );
    }

    private void handleTokenResponse(Response<AccessTokenResponse> tokenResponse) {
        if (tokenResponse.isSuccessful() && tokenResponse.code() == 200) {
            mSharedPrefUtils.saveAccessToken(Objects.requireNonNull(tokenResponse.body()).getAccessToken().getAccessToken());
            mSharedPrefUtils.saveOldToken(Objects.requireNonNull(tokenResponse.body()).getAccessToken().getToken());
            ifViewAttached(SplashView::onTokenFetched);
        } else
            ifViewAttached(view -> view.onTokenError(tokenResponse.message()));
    }
}

File: app/src/main/java/omniex/nl/omniex/ui/app/splash/SplashPresenter.kt
package omniex.nl.omniex.ui.app.splash

import omniex.nl.omniex.data.model.auth.OldToken
import omniex.nl.omniex.data.model.response.AccessTokenResponse
import omniex.nl.omniex.data.repository.TokenRepository
import omniex.nl.omniex.ui.base.BasePresenter
import omniex.nl.omniex.utils.SharedPrefUtils
import retrofit2.Response
import timber.log.Timber
import javax.inject.Inject

class SplashPresenter @Inject
internal constructor(private val mTokenRepository: TokenRepository, private val mSharedPrefUtils: SharedPrefUtils) : BasePresenter<SplashView>() {

    internal fun resolveAccessToken() {
        if (mSharedPrefUtils.accessToken().isEmpty())
            getAccessToken()
        else
            refreshAccessToken()
    }

    private fun getAccessToken() {
        addDisposable(mTokenRepository
                .accessToken()
                .subscribe(
                        { this.handleTokenResponse(it) },
        { error ->
            ifViewAttached { view -> view.onTokenError(error.message!!) }
            Timber.e(error)
        })
        )
    }

    private fun refreshAccessToken() {
        addDisposable(mTokenRepository
                .refreshAccessToken(OldToken(mSharedPrefUtils.oldToken()))
                .subscribe(
                        { this.handleTokenResponse(it) },
        { error ->
            ifViewAttached { view -> view.onTokenError(error.message!!) }
            Timber.e(error)
        })
        )
    }

    private fun handleTokenResponse(tokenResponse: Response<AccessTokenResponse>) {
        if (tokenResponse.isSuccessful() && tokenResponse.code() == 200) {
            mSharedPrefUtils.saveAccessToken(tokenResponse.body()!!.accessToken!!.accessToken)
            mSharedPrefUtils.saveOldToken(tokenResponse.body()!!.accessToken!!.token!!)
            ifViewAttached({ it.onTokenFetched() })
        } else
            ifViewAttached { view -> view.onTokenError(tokenResponse.message()) }
    }
}
